graph.cpp

```diff
@@ -0,0 +1,194 @@
+#include "graph.hpp"
+#include<vector>
+#include<iostream>
+#include<queue>
+
+void Graph::Node::insertIncoming(int64 from) {
+  incoming->insert(from);
+}
+
+void Graph::Node::insertOutgoing(int64 to) {
+  outgoing->insert(to);
+}
+
+void Graph::Node::eraseIncoming(int64 from) {
+  incoming->erase(from);
+}
+
+void Graph::Node::eraseOutgoing(int64 to) {
+  outgoing->erase(to);
+}
+
+bool Graph::Node::containsEdgeTo(int64 to) {
+  // apparently the standard c++ way to check if something is in range
+  // call this too many times to not make this a method...
+  return outgoing->find(to) != outgoing->end();
+}
+
+bool Graph::Node::containsEdgeFrom(int64 from) {
+  return incoming->find(from) != incoming->end();
+}
+
+bool Graph::Contains(int64 nodeid) {
+  return nodemap.find(nodeid) != nodemap.end();
+}
+
+Graph::Node Graph::Node::DeepCopy() {
+  // create a new node
+  auto result = Node();
```

Comment:
Some people at Google don't mind auto. I hate it because most of the time I encounter auto, I have to jump through a few links in order to figure out wtf the type is. In general, because code is read much more often than it's written, the use of auto is discouraged unless it's really blaringly obvious what the thing is. My policy is that because it's kindof annoying to make a judgement call every time I use auto, I just avoid it.